clang/lib/Frontend/CASDependencyCollector.cpp

```diff
@@ -0,0 +1,71 @@
+//===--- CASDependencyCollector.cpp ---------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "clang/Frontend/CASDependencyCollector.h"
+#include "clang/Basic/DiagnosticCAS.h"
+#include "llvm/CAS/CASDB.h"
+#include "llvm/CAS/CASOutputBackend.h"
+#include "llvm/Support/VirtualOutputBackends.h"
+
+using namespace clang;
+using namespace clang::cas;
+
+CASDependencyCollector::CASDependencyCollector(
+    const DependencyOutputOptions &Opts,
+    IntrusiveRefCntPtr<llvm::cas::CASOutputBackend> OutputBackend)
+    : DependencyFileGenerator(Opts, llvm::vfs::makeNullOutputBackend()),
+      CASOutputs(std::move(OutputBackend)), OutputName(Opts.OutputFile) {}
+
+llvm::Error CASDependencyCollector::replay(const DependencyOutputOptions &Opts,
+                                           CASDB &CAS, ObjectRef DepsRef,
+                                           llvm::raw_ostream &OS) {
+  auto Refs = CAS.load(DepsRef);
+  if (!Refs)
+    return Refs.takeError();
+
+  CASDependencyCollector DC(Opts, nullptr);
+  auto Err = CAS.forEachRef(*Refs, [&](ObjectRef Ref) -> llvm::Error {
+    auto PathHandle = CAS.load(Ref);
+    if (!PathHandle)
+      return PathHandle.takeError();
+    StringRef Path = CAS.getDataString(*PathHandle);
+    // This assumes the replay has the same filtering options as when it was
+    // originally computed. That avoids needing to store many unnecessary paths.
+    DC.addDependency(Path);
```

Comment:
Small nitpick, I'd suggest adding a FIXME that, if path remapping is enabled, we should be remapping the path from canonical to the invocation's environment.

Reply:
Done